### PR TITLE
Row-first option for columnize

### PR DESCRIPTION
**Pull request by naught101**
This adds a row-first sort to the columnize function, to address part of #8741 .

Original columnise: 

``` python
In [9]: print(columnize(items))
a  c  eeeeeeeeeeeeeeeeeeeeeee  eeeeeeeeeeeeeeeeeeeeeee
b  d  eeeeeeeeeeeeeeeeeeeeeee  eeeeeeeeeeeeeeeeeeeeeee
```

New addition:

``` python
In [10]: print(columnize(items, row_first=True))
a                        b                        c                      
d                        eeeeeeeeeeeeeeeeeeeeeee  eeeeeeeeeeeeeeeeeeeeeee
eeeeeeeeeeeeeeeeeeeeeee  eeeeeeeeeeeeeeeeeeeeeee
```

### CODE REVIEW

File: IPython/utils/text.py [hunk 1]
```diff
@@ -714,6 +725,10 @@ def columnize(items, separator='  ', displaywidth=80):
     items : sequence of strings
         The strings to process.
 
+    row_first : (default False)
+        Whether to to compute columns for a row-first matrix instead ofr
```
Comment (Carreau) [inline IPython/utils/text.py]: instead `ofr` :-)

**Comment by Carreau** [comment]
Looks good to me. 

Want to add test for the option in `IPython/utils/tests/test_text.py` ?

File: IPython/utils/text.py (hunk 1 above)
Comment (naught101) [inline IPython/utils/text.py]: also, 'to to'.. gah..

**Comment by naught101** [comment]
Guess that was a good idea :)

I think it's ready to go.

**Comment by Carreau** [comment]
> Guess that was a good idea :)

:-)

**Comment by naught101** [comment]
Hrm.. thoughts please:

currently, the row-first option calculates the optimal layout by starting with the max number of rows, then calculating the minimum number of columns, which may end up being much smaller than the display width. For example, this is the current result:

```{python}
In [3]: [s*15 for s in 'abcdef']
Out[3]: 
['aaaaaaaaaaaaaaa', 'bbbbbbbbbbbbbbb', 'ccccccccccccccc'
 'ddddddddddddddd', 'eeeeeeeeeeeeeee', 'fffffffffffffff']
```

Another option would be to maximize the columns, while remaining under the display limit. which would look like this:

```{python}
In [3]: [s*15 for s in 'abcdef']
Out[3]: 
['aaaaaaaaaaaaaaa', 'bbbbbbbbbbbbbbb', 'ccccccccccccccc', 'ddddddddddddddd'
 'eeeeeeeeeeeeeee', 'fffffffffffffff']
```

which might look nicer in some cases. However, it would never be more efficient in terms of space used, and it would require somewhat more complex code. 

Do you care which option is used?

**Comment by Carreau** [comment]
I think the point of solution 1 was to make completion always closer to the cursor if you wrap around, and as the qtconsole is now it's own package, columnize should  not even used anymore (i grepped quickly but you can re-check)... so feel free to refactor any way you want.

**Comment by Carreau** [comment]
We can also merge that and you can update the algo in another PR.

**Comment by naught101** [comment]
Ok, I just had a though, and added a 'spread' option, which makes columnize spread the columns out to the maximum allowable width (optimal spacing was already calculated, I just left justified the separator to match the padding). 4 Tests included.

As far as I am concerned, this is ready to go - good enough for the pretty print issue anyway.

**Comment by Carreau** [comment]
Ok, let's move forward then

**Pull request merged**